llvm/lib/Target/AIE/ReservedRegsLICM.cpp

```diff
@@ -0,0 +1,381 @@
+//
+// This pass performs loop invariant code motion on machine instructions that
+// define reserved physical registers. It can move those to the preheader or
+// the exit block. This is based on the LivePhysRegs analysis and tracking of
+// regs that are altered within loops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIE.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/CodeGen/LivePhysRegs.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineLoopInfo.h"
+#include "llvm/CodeGen/MachineRegisterInfo.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/Debug.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "reserved-reg-licm"
+
+namespace {
+
+/// Track registers that have been defined/changed
+class RegDefMap {
+  const TargetRegisterInfo &TRI;
+  DenseMap<MCRegister, MachineInstr *> UniqueDefs;
+  BitVector PhysRegChanged;
+
+public:
+  RegDefMap(const TargetRegisterInfo &TRI)
+      : TRI(TRI), PhysRegChanged(TRI.getNumRegs()) {}
+
+  /// Track every register that was changed by \p MI
+  void addChangedRegs(MachineInstr &MI);
+
+  /// Whether \p Reg or any of its aliases has been changed.
+  bool hasChanged(MCRegister Reg) const;
+
+  /// Whether \p Reg has been defined a single time.
+  MachineInstr *getUniqueDef(MCRegister Reg) const;
+};
+
+void RegDefMap::addChangedRegs(MachineInstr &MI) {
+  for (const MachineOperand &MO : MI.operands()) {
+    if (MO.isRegMask()) {
+      PhysRegChanged.setBitsNotInMask(MO.getRegMask());
+      UniqueDefs.clear();
+    } else if (MO.isReg() && MO.isDef() && MO.getReg().isPhysical()) {
+      MCRegister Reg = MO.getReg();
+      if (!PhysRegChanged.test(Reg)) {
+        assert(!UniqueDefs.contains(Reg));
+        UniqueDefs[Reg] = &MI;
+      } else {
+        UniqueDefs.erase(Reg);
+      }
+      PhysRegChanged.set(Reg);
+    }
+  }
+}
+
+bool RegDefMap::hasChanged(MCRegister Reg) const {
+  assert(range_size(TRI.regunits(Reg)) == 1 && "Phys reg has aliases.");
+  return PhysRegChanged.test(Reg);
+}
+
+MachineInstr *RegDefMap::getUniqueDef(MCRegister Reg) const {
+  assert(range_size(TRI.regunits(Reg)) == 1 && "Phys reg has aliases.");
+  if (auto It = UniqueDefs.find(Reg); It != UniqueDefs.end())
+    return It->second;
+  return nullptr;
+}
+
+class ReservedRegsLICM : public MachineFunctionPass {
+  const TargetRegisterInfo *TRI;
+  MachineRegisterInfo *MRI;
+
+  /// Whether the MachineFunction got changed
+  bool Changed = false;
+
+public:
+  static char ID;
+  ReservedRegsLICM() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.addRequired<MachineLoopInfo>();
+    AU.addPreserved<MachineLoopInfo>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+private:
+  /// Keep track of information about hoisting candidates.
+  struct CandidateInfo {
```

Comment:
nit: you can encapsulate all Candidate-related stuff in class. Something like this:

```
  struct Candidates {
    struct CandidateInfo {
      CandidateInfo(MCPhysReg DefinedReg) : DefinedReg(DefinedReg) {}
      MCPhysReg DefinedReg = MCRegister::NoRegister;
      MachineInstr *HoistCandidate = nullptr;
    };
    const MachineRegisterInfo* MRI;
    DenseMap<MCPhysReg, CandidateInfo> AllCandidates;
    Candidates(const MachineRegisterInfo* MRI) : MRI(MRI) {}
    CandidateInfo *getCandidateInfo(const MachineInstr &MI);
  };
  using CandidateInfo = Candidates::CandidateInfo;
```
Also, you can use this as a local object in Hoist/Sink methods, not as an attribute that you need to clean before the second use....

Reply:
Agreed, that is always better to avoid keeping too much state ;)